lib/tapioca/gem/listeners/foreign_constants.rb

```diff
@@ -35,10 +35,10 @@ def on_scope(event)
               # base constant. Then, generate RBIs as if the base constant is extending the mixin,
               # which is functionally equivalent to including or prepending to the singleton class.
               if !name && constant.singleton_class?
-                name = constant_name_from_singleton_class(constant)
-                next unless name
+                attached_class = attached_class_of(constant)
```

Comment:
Another remark:

Looks like singleton classes already have a pointer to their attached classes in underlying implementation (at least in MRI). Which is like, obvious when I think about it (d'uh, otherwise how could they generate the `::inspect`?).

It's currently stored in an ivar called `id__attached` (the constant in the C code is `id__attached__`):

https://github.com/ruby/ruby/blob/4325e90205aa4cd0ea031df1b5e6334bfd9c7e51/class.c#L33

Usage:

https://github.com/ruby/ruby/blob/ff07e5c264c82f73b0368dd0bc2ae39f78678519/object.c#L1542

Perhaps we could pitch making a public API that exposes this, instead of using this Stringy hack to begin with? (long term, this PR can still proceed as normal)

Reply:
I can try to make another case on the issue with reflection/introspection as the main use case.

Reply:
I made a case for it here: https://bugs.ruby-lang.org/issues/12084#note-11